concordium-contracts-common/src/schema_json.rs

```diff
@@ -956,13 +956,305 @@ impl Fields {
+    // defined by the last argument (`3$` == `indent`). This creates the
+    // correct tabs/whitespace. https://doc.rust-lang.org/std/fmt/#fillalignment
+    out = format!("{}{:>3$}{}:\n", out, "", type_schema_name, indent);
+    out = format!(
+        "{}{:>3$}{}\n",
+        out,
+        "",
+        type_schema.to_string().replace('\n', "\n        "),
+        indent + 2
+    );
+    out
+}
+
+/// Displays a pretty-printed template of the [`VersionedModuleSchema`].
+///
+/// # Examples
+///
+/// ## Display a template of the `VersionedModuleSchema`
+///
+/// ```
+/// # use serde_json::json;
+/// # use concordium_contracts_common::schema_json::*;
+/// # use concordium_contracts_common::schema::*;
+/// # use concordium_contracts_common::*;
+/// # use std::collections::BTreeMap;
+/// #
+/// let mut receive_function_map = BTreeMap::new();
+/// receive_function_map.insert(String::from("MyFunction"), FunctionV2 {
+///     parameter:    Some(Type::AccountAddress),
+///     error:        Some(Type::AccountAddress),
+///     return_value: Some(Type::AccountAddress),
+/// });
+///
+/// let mut map = BTreeMap::new();
+///
+/// map.insert(String::from("MyContract"), ContractV3 {
+///     init:    Some(FunctionV2 {
+///         parameter:    Some(Type::AccountAddress),
+///         error:        Some(Type::AccountAddress),
+///         return_value: Some(Type::AccountAddress),
+///     }),
+///     receive: receive_function_map,
+///     event:   Some(Type::AccountAddress),
+/// });
+///
+/// let schema = VersionedModuleSchema::V3(ModuleV3 {
+///     contracts: map,
+/// });
+///
+/// let display = "Contract:                     MyContract
+///   Init:
+///     Parameter:
+///       \"<AccountAddress>\"
+///     Error:
+///       \"<AccountAddress>\"
+///     Return value:
+///       \"<AccountAddress>\"
+///   Methods:
+///     - \"MyFunction\"
+///       Parameter:
+///         \"<AccountAddress>\"
+///       Error:
+///         \"<AccountAddress>\"
+///       Return value:
+///         \"<AccountAddress>\"
+///   Event:
+///     \"<AccountAddress>\"\n";
+///
+/// assert_eq!(display, format!("{}", schema));
+/// ```
+impl Display for VersionedModuleSchema {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        let mut out = String::new();
+        match self {
+            VersionedModuleSchema::V0(module_v0) => {
+                for (contract_name, contract_schema) in module_v0.contracts.iter() {
+                    out = format!("Contract: {:>30}\n", contract_name);
+                    // State
+                    if let Some(type_schema) = &contract_schema.state {
+                        out = format!("{}{:>2}State:\n", out, "");
+                        out = format!(
+                            "{}{:>4}{}\n",
+                            out,
+                            "",
+                            type_schema.to_string().replace('\n', "\n    ")
```

Comment:
I usually don't like this sort of formatting, and would prefer if we could use a crate like https://crates.io/crates/indenter for indenting data, which is also more efficient than converting to string and then replacing things.

But with the work that is already done, changing to using that is going to be quite a bit of effort, so if you've sufficiently tested this let's just go ahead and merge.

Reply:
Good point. I will explore that crate and try to use it next time. For now, I left the current way of formatting the string.